### PR TITLE
[WIP] Corrections Schema

### PR DESCRIPTION
**Pull request by nwinner**
We need a method for dealing with the compatibility of different defect corrections. i.e. the corrected energy should not include multiple electrostatic corrections or multiple potential alignment corrections.

The best way to do deal with this, imo, is to have new correction objects that handle it. The current `DefectEntry` hard codes "freysoldt" for getting corrections, but one might want to use eFNV or the 2D-FNV correction.

Current idea is to define each correction with a `Correction` object that holds energy and metadata, and then a `DefectEntry` holds a `CorrectionSummary` which holds multiple corrections, mapped with a dictionary of `CorrectionType`. This way, if you try to add, say, a eFNV correction, it will replace the FNV correction in the dict.

### CODE REVIEW

**Comment by nwinner** [comment]
@jmmshn How do you feel about using a schema like this for corrections?

If you like this solution I can fix the workflow failures and we can iterate on it.

**Comment by jmmshn** [comment]
Hi @nwinner, is there a specific reason that we are splitting the `electrostatic` and `potential_alignment` parts like this?
Is the plan to allow people to use the anisotropic electrostatic correction regardless of if the `LOCPOT` is available or not?

**Comment by nwinner** [comment]
As I see it there are a few reasons to do it this way.

1. Some corrections do not have a potential alignment part. Say we want to add Madelung or Makov Payne corrections (slightly older, maybe for comparison), which don't have a potential alignment part. It would make sense, to me, to separate it to avoid confusion.
2. Like you said, if a LOCPOT is not available then you want to know that the pot alignment was not done to avoid double counting down the line.
3. Better for interpretability for new users.
4. Right now we just have electrostatic+pot alignment, but down the line you might want to add a host of other corrections like elastic, magnetic, or band filling corrections. At that point we'll need a way to separate the different types so might as well start here.

On the down side I can see one thing; the above method cannot check if the pot alignment and electrostatic correction come from the same scheme. i.e. this leaves open something like Fresyoldt electrostatic + Kumagai pot alignment. I don't really know if that would cause issues, but its something to note. It won't be an issue, though, unless someone manually puts the two together, since the functions you have written will ensure the same scheme is used for both.  

Open to suggestions if you think its unnecessarily complicated.

**Comment by jmmshn** [comment]
>  1. Some corrections do not have a potential alignment part. Say we want to add Madelung or Makov Payne corrections (slightly older, maybe for comparison), which don't have a potential alignment part. It would make sense, to me, to separate it to avoid confusion.

It seems to me that if the potential alignment is a concept that is not universal to all correction scheme then we should not have it sit at top level.  So each correction scheme should only keep track of the different parts of the correction internally and report the final correction value.

> 3. Better for interpretability for new users.

I think this only affects new users coming from pyCDT.  I think it's likely that most users will not think too much about what the correction schemes are doing and even less about different parts of the correction schemes.  

I think the best way for now is to have a running enum of the different correction schemes and just have each implementation keep track of the different parts internally.

> 4. Right now we just have electrostatic+pot alignment, but down the line you might want to add a host of other corrections like elastic, magnetic, or band filling corrections. At that point we'll need a way to separate the different types so might as well start here.

I think if someone wants to compose correction schemes like that then they will already likely know what they are doing and it will be pretty easy for them to just come up with a new name in the enum like `FreySoldt_w_band_filling` and simply have it call both function and add it up.

**Comment by nwinner** [comment]
Hmm I can't say I agree with the idea of adding `FreySoldt_w_band_filling` as a potential future solution, but I'll defer to you on keeping electrostatic and potential alignment together for now.

So you don't want a dictionary or list of corrections in the `DefectEntry` object? Just a `Correction` object with one energy value? 

If that's so I'll refactor

**Comment by jmmshn** [comment]
> Hmm I can't say I agree with the idea of adding FreySoldt_w_band_filling as a potential future solution, but I'll defer to you on keeping electrostatic and potential alignment together for now.

OK bad example but maybe `SteveCorrection`. The point is that the additional engineering around how to break different corrections up should be left to the individual implementaion as long as the individual implmentations are not self-contained it should be easy to compose them into w/e combo correction scheme as long as the combo correction scheme is codified and named we will know exactly what was done and how to reproduce.

File: pymatgen/analysis/defects/corrections/base.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+from dataclasses import dataclass
+from enum import Enum
+from typing import Any, Optional
+
+import numpy as np
+from monty.json import MSONable
+
+
+class CorrectionType(Enum):
```
[the remark this answers is not in the document]
Comment (nwinner) [reply pymatgen/analysis/defects/corrections/base.py:11]: I could see someone being interested in knowing the relative contributions of each as part of their work, but probably debugging is the only common use for it.

**Comment by jmmshn** [comment]
@nwinner I'm making some much needed cleanup for to `thermo.py` only this one should be relevant to this PR:
`correction_metadata`  -> `corrections_metadata`

**Comment by jmmshn** [comment]
ok a bunch of random linting stuff broke due to upstream changes to pre-commit 

https://stackoverflow.com/questions/75269700/pre-commit-fails-to-install-isort-5-11-4-with-error-runtimeerror-the-poetry-co

Should be all fixed now on `main` now.

**Pull request closed without merging**